### DISCUSSION TITLE
Extremly Complicated gui, learning pyqt, and qt just to contribute

### DISCUSSION THREAD

**Discussion by Bryfour** [FreeCAD/FreeCAD#9665] [Code contribution discussions]
I'm looking to change the gui to make it easier for cad users to move to. I've used fusion since i was abut 9, and I've used Solidworks for 3 years in school. I saw Thomas Sanlanders yt vid on Free CAD and thought he had some valid points. I'm learning to code just to contribute, I'm hoping to contribute first by changing the gui, and making things easier to find. Sketch and solid modeling should be in the same menu, and the buttons are tiny. I can make art for the buttons if I need to, I have not been able to find the dimensions tool yet. 
Ill make the buttons bigger first, and re organize the menus as is right now, hopefully i can help fix bugs in he future. I really appreciate this being free and open source. 
Thoughts?

**Comment by chennes** [comment]
Welcome, @Bryfour -- you'll probably want to consider setting up an account at https://forum.freecad.org -- many users and developers discuss user interface issues there. Of course, you can also drop in to our Discord, IRC, etc. -- you'll find all those links at the top of the homepage at https://freecad.org. Because you are proposing changes to something that will affect basically all users you will definitely want to have others look at it and discuss it before you spend too much time coding, etc.

**Reply by Bryfour** [reply]
Well, im looking at this as a learning experience, but i will probably find a way to make this go through, whether its with small changes over time, or just a big change. I hope to make it easier to navigate. thanks.

**Reply by chennes** [reply]
You will 100% want to make small changes -- we ask that even developers who have been around for years keep their PRs as small as possible. Think ~100 lines as a good target for ensuring a timely code review.

**Reply by Bryfour** [reply]
Maximum 100 lines?
From of what ive understood based on the framework, the gui is in python, and i hope that mans im only going to need to make value changes, like button sizes etc.

**Reply by Pesc0** [reply]
Do not see it as a hard maximum, it's just that bigger changes require more careful review and thus more time by reviewers. Also changing a lot of stuff at once carries greater possibilities for bugs and large changes are probably more controversial in general.

**Reply by Bryfour** [reply]
Yes, since im learning, thats the thought process. Also i kinda figured that any big changes would make people unhappy. Gotta boil the frog in ccold water... in a good way.

**Reply by Bryfour** [reply]
For anyone who sees this post, I want to know your honest opinion about what could be changed about the Ui to make it easier/ fewer steps to locate and use different tools/ features.
Here is my thought:
1 Merge the sketch dropdown and part dropdown, and create a left to right sliding menu, that's either for solid parts or sketches
2 Add the key "L" in sketching, 1 press = Line, 2 Press = Construction line
3 Increase the button size to be 3x current size but only one Toolbar worth of buttons. Reason (I'm used to finding buttons based on the image)
4 Combine Vertical and horizontal constraint into one button and change the image to something like : ll=


Thoughts from other people: 
?

**Reply by Bryfour** [reply]
@chennes I've been thinking. I think at the moment, even reorganizing the default layout would be helpful. I noticed in the forums that a new user wizard is being made, so yeah.

**Comment by luzpaz** [comment]
FreeCAD has many users who hail from different proprietary CAD suites. Many of them want to change FreeCAD UI to reflect the workflows they are used to. As you can imagine there are a lot of different passionate opinions on this topic. A good person to follow is @PaddleStroke who came in and was able to get a lot of GUI discussions (and ultimately) core changes implemented.

**Reply by Bryfour** [reply]
Nice, Maybe I ll learn something.

**Comment by Pesc0** [comment]
I too love the fusion interface: I find it visually appealing, but most importantly easy to learn and navigate. I had expressed some thoughts and ideas in this comment: https://github.com/FreeCAD/FreeCAD/issues/8815#issuecomment-1464946263 
How I've customized my interface (fusion-like, no coding required):

![image](https://github.com/FreeCAD/FreeCAD/assets/54720203/9caf402c-5d98-42e4-93da-42ec1ce6fc81)

See also this (now unmaintained) addon: https://github.com/HakanSeven12/Modern-UI

List of other UI related issues: https://github.com/FreeCAD/FreeCAD/labels/UI%2FUX

Roadmap: https://freecad.github.io/DevelopersHandbook/roadmap/

> I have not been able to find the dimensions tool yet

https://github.com/FreeCAD/FreeCAD/issues/8561

With that said: welcome, I hope you'll have a good journey :D

**Reply by Bryfour** [reply]
@Pesc0  Is there a way for you to export that layout so i can use it as well?

**Reply by Pesc0** [reply]
I would encourage you to play around with it yourself. It's very easy: you can create custom toolbars in Tools > Customize > Toolbars. Then you just move stuff around. Or search for "freecad customize toolbars" on youtube.
Maybe you can make it even better, or at least get a better idea of what you want to achieve ;)

**Comment by luzpaz** [comment]
It's also a great opportunity to help us expand https://github.com/FreeCAD/DevelopersHandbook/ (if you choose to) while you learn how to contribute to FreeCAD. 

JFYI, better to learn pyside6 and shiboken, not pyqt.

**Reply by Bryfour** [reply]
I thought that the framework was in qt, I read something about it in documentation

**Reply by adrianinsaval** [reply]
pyside is qt's official python interface. pyqt is unofficial similar implementation. FreeCAD uses pyside, using pyqt is probably possible but not advised

**Comment by PaddleStroke** [comment]
Hi Bryfour,
Welcome abord! 
So I've been at this place about 1.5 year ago, so here's some advices to keep you from failing in your quest.
- Don't try to overhaul the UI at once as it will never happen. Even very small UI changes can be very tedious to get accepted.
- Smaller PR tend to merge faster, as Chris said, 100 lines or less is best if you want something to merge. However sometimes 100 lines won't cut it. So if reworking a widget or a task panel or what not, needs 3000 lines, then you can still do it but just make sure that the changes have been broadly accepted or that it solves an existing issue. Try to get a clear commit history (yeah as if I was doing it, it's harder than it sounds). And be prepared to be patient.
- Before starting a project, make a mockup, post it on the forum and see what people say. For instance increasing the toolbar icon size, I'm pretty sure that people are going to be mostly against it. So before you start it you need to plan and find an acceptable solution.
- For the toolbar, if you want to change them to look like f360, even if you make your PR perfect, it won't merge. The reason is that toolbar organization depends a lot on individual. So I can tell you that a LOT of people on the forum are going to be against such a change and that you'll never hear the end of it. So instead what you want to do, is to implement your toolbar organization as an option. Such that no one is going to be against it. So you can just make a preference setting combobox "Interface" with several options "Standard" and "Modern" for example. And you can implement your solution in "Modern" and leave the current default ui in "Standard". Then maybe in the future, when more people will come to use your UI, maybe the default UI will change to yours and the old one will disappear.
- You need to learn c++ if you want to do any core changes. Because the core of FreeCAD is still writen in c++, so if you want to implement a "modern UI", you probably want to do it in the core in c++. The other option is to make an addon in python, exactly what [HakanSeven12/Modern-UI](https://github.com/HakanSeven12/Modern-UI) has done. The issue is that not many people use addons and that you need to maintain. So most likely your changes will get lost in time.

So yes if you want to bring changes to the UI, I suggest you do something similar to [HakanSeven12/Modern-UI](https://github.com/HakanSeven12/Modern-UI) but in the core in cpp, and as an option.

Having said that, I would suggest starting with something smaller. You can try to find smaller things by looking at things that makes no sense in the UI. Or you can also have a look at the Linkstage branch of Realthunder. It is known to have a lot of UI improvements and quality of life features that have not merged into master. So you could try to cherry pick some of these and make PR for them.
One example of such : Realthunder added settings to change the size of the Tree text and icon size.

**Reply by nielsnl68** [reply]
Hi, i'm relative new user of FreeCAD, i used a couple of CAD software but now i am trying to switch to FreeCAD,
Because @Bryfour asked for some suggestions to get be changed, i will point them below.
Please be aware that i know that this point should be posted at a different place (and i will do that as well).

- One GUI issue i really dislike is when you use Parametric Spreadsheet's to setup a sketch, that you need to go to 2 dialog's to set the parameters. It would be nice when you can do that direct in the first dialog screen.

- An less GUI related issue but still very annoying is that when you have open DoF's you are in the dark which constraints are missing. It would be nice when you hoover over the number of DoF's that you get a list with the open constraints.

Other then that i'm starting to see the benefit using FreeCAD,

**Reply by Bryfour** [reply]
@PaddleStroke  Thanks for your input. I t makes total sense to make the UI have a Layout Named Old style, and a new layout called Modern. When I mentioned making the buttons larger, i would only have one taskbar ribbon to keep it the same size ,but show and hide buttons based on what you are doing. 
I'm wondering how I'm supposed to find all the stuff on v code, i downloaded it, and I don't know where I go to find the UI info. Is there a page in the devs guide that would help? 

Also, the UI is in QT from C++, so that's what I'm learning to start with.

**Comment by nielsnl68** [comment]
Hi @PaddleStroke.

Your point's are so true. And not only for this project but for most projects on GitHub.

**Comment by mfro0** [comment]
Icon size can be adjusted in preferences
